src/service/parallax-impl.js

```diff
@@ -0,0 +1,215 @@
+    const previousScrollTop = this.getPreviousScroll_();
+    const delta = previousScrollTop - newScrollTop;
+
+    elements.forEach(element => {
+      if (!element.shouldUpdate(viewport)) {
+        return;
+      }
+      element.update(delta);
+      this.setPreviousScroll_(newScrollTop);
+    });
+
+    this.fire_(newScrollTop);
+  }
+
+  /**
+   * Create a value for the CSS transform property given a position.
+   * @param {number} position
+   * @return {string}
+   */
+  transform_(position) {
+    return `translate3d(0,${position.toFixed(2)}px,0)`;
+  }
+
+  /**
+   * Get the previous scroll value.
+   * @return {number}
+   * @private
+   */
+  getPreviousScroll_() {
+    return this.previousScroll_;
+  }
+
+  /**
+   * Set the previous scroll value.
+   * @param {number} scroll
+   * @private
+   */
+  setPreviousScroll_(scroll) {
+    this.previousScroll_ = scroll;
+  }
+
+  /**
+   * Add listeners to parallax scroll events.
+   * @param {!function()} cb
+   * @private
+   * @visibleForTesting
+   */
+  addScrollListener_(cb) {
+    this.parallaxObservable_.add(cb);
+  }
+
+  /**
+   * Remove listeners from parallax scroll events.
+   * @param {!function()} cb
+   * @private
+   * @visibleForTesting
+   */
+  removeScrollListener_(cb) {
+    this.parallaxObservable_.remove(cb);
+  }
+
+  /**
+   * Alert listeners that a scroll has occurred.
+   * @param {number} scrollTop
+   * @private
+   */
+  fire_(scrollTop) {
+    this.parallaxObservable_.fire(scrollTop);
+  }
+}
+
+/**
+ * Encapsulates and tracks an element's linear parallax effect.
+ */
+export class ParallaxElement {
+  /**
+   * @param {!Element} element The element to give a parallax effect.
+   * @param {!function(number):string} transform Computes the transform from the position.
+   */
+  constructor(element, transform) {
+    const factor = element.getAttribute(ATTR);
+
+    /** @private @const {!Element} */
+    this.element_ = element;
+
+    /** @private @const {!function(number):string} */
+    this.transform_ = transform;
+
+    /** @private @const {number} */
+    this.factor_ = (factor ? parseFloat(factor) : 0.5) - 1;
+
+    /** @private {number} */
+    this.offset_ = 0;
+  }
+
+  /**
+   * Apply the parallax effect to the offset given how much the page
+   * has moved since the last frame.
+   * @param {number} delta The movement of the base layer e.g. the page.
+   */
+  update(delta) {
+    this.offset_ += delta * this.factor_;
+    setStyles(this.element_, {transform: this.transform_(this.offset_)});
+  }
+
+  /**
+   * True if the element is in the viewport.
+   * @param {!./viewport-impl.Viewport} viewport
+   * @return {boolean}
+   */
+  shouldUpdate(viewport) {
+    const viewportRect = viewport.getRect();
+    const elementRect = viewport.getLayoutRect(this.element_);
+    elementRect.top -= viewportRect.top;
+    elementRect.bottom = elementRect.top + elementRect.height;
+    return this.isRectInView_(elementRect, viewportRect.height);
```

Comment:
We have a [helper function](https://github.com/ampproject/amphtml/blob/93b67b0/src/layout-rect.js#L83-L92) that does exactly this.